go/consensus/tendermint/apps/registry/state/state.go

```diff
@@ -552,123 +572,156 @@ func NewImmutableState(state abci.ApplicationState, version int64) (*ImmutableSt
+	if err := s.Tree.Remove(ctx, signedNodeByEntityKeyFmt.Encode(&node.EntityID, &node.ID)); err != nil {
+		return abciAPI.UnavailableStateError(err)
+	}
+	if err := s.Tree.Remove(ctx, nodeStatusKeyFmt.Encode(&node.ID)); err != nil {
+		return abciAPI.UnavailableStateError(err)
+	}
 
 	address := []byte(tmcrypto.PublicKeyToTendermint(&node.Consensus.ID).Address())
-	s.tree.Remove(nodeByConsAddressKeyFmt.Encode(address))
+	if err := s.Tree.Remove(ctx, nodeByConsAddressKeyFmt.Encode(address)); err != nil {
+		return abciAPI.UnavailableStateError(err)
+	}
 
-	s.tree.Remove(keyMapKeyFmt.Encode(&node.Consensus.ID))
-	s.tree.Remove(keyMapKeyFmt.Encode(&node.P2P.ID))
+	if err := s.Tree.Remove(ctx, keyMapKeyFmt.Encode(&node.Consensus.ID)); err != nil {
+		return abciAPI.UnavailableStateError(err)
+	}
+	if err := s.Tree.Remove(ctx, keyMapKeyFmt.Encode(&node.P2P.ID)); err != nil {
+		return abciAPI.UnavailableStateError(err)
+	}
 
 	certHash := nodeCertificateToMapKey(node.Committee.Certificate)
-	s.tree.Remove(certificateMapKeyFmt.Encode(&certHash))
+	if err := s.Tree.Remove(ctx, certificateMapKeyFmt.Encode(&certHash)); err != nil {
+		return abciAPI.UnavailableStateError(err)
+	}
+
+	return nil
 }
 
-func (s *MutableState) SetRuntime(rt *registry.Runtime, sigRt *registry.SignedRuntime, suspended bool) error {
-	ent, err := s.getSignedEntityRaw(rt.EntityID)
-	if ent == nil || err != nil {
-		return registry.ErrNoSuchEntity
+// SetRuntime sets a signed runtime descriptor for a registered runtime.
+func (s *MutableState) SetRuntime(ctx context.Context, rt *registry.Runtime, sigRt *registry.SignedRuntime, suspended bool) error {
+	if err := s.Tree.Insert(ctx, signedRuntimeByEntityKeyFmt.Encode(&rt.EntityID, &rt.ID), []byte("")); err != nil {
+		return abciAPI.UnavailableStateError(err)
 	}
 
-	s.tree.Set(signedRuntimeByEntityKeyFmt.Encode(&rt.EntityID, &rt.ID), []byte(""))
-
+	var err error
 	if suspended {
-		s.tree.Set(suspendedRuntimeKeyFmt.Encode(&rt.ID), cbor.Marshal(sigRt))
+		err = s.Tree.Insert(ctx, suspendedRuntimeKeyFmt.Encode(&rt.ID), cbor.Marshal(sigRt))
 	} else {
-		s.tree.Set(signedRuntimeKeyFmt.Encode(&rt.ID), cbor.Marshal(sigRt))
+		err = s.Tree.Insert(ctx, signedRuntimeKeyFmt.Encode(&rt.ID), cbor.Marshal(sigRt))
 	}
-
-	return nil
+	return abciAPI.UnavailableStateError(err)
 }
 
-func (s *MutableState) SuspendRuntime(id common.Namespace) error {
-	_, raw := s.Snapshot.Get(signedRuntimeKeyFmt.Encode(&id))
-	if raw == nil {
+// SuspendRuntime marks a runtime as suspended.
+func (s *MutableState) SuspendRuntime(ctx context.Context, id common.Namespace) error {
+	data, err := s.Tree.RemoveExisting(ctx, signedRuntimeKeyFmt.Encode(&id))
+	if err != nil {
+		return abciAPI.UnavailableStateError(err)
+	}
+	if data == nil {
 		return registry.ErrNoSuchRuntime
 	}
-
-	s.tree.Remove(signedRuntimeKeyFmt.Encode(&id))
-	s.tree.Set(suspendedRuntimeKeyFmt.Encode(&id), raw)
-	return nil
+	return s.Tree.Insert(ctx, suspendedRuntimeKeyFmt.Encode(&id), data)
```

Comment:
ought this one be wrapped in UnavailableStateError?

Reply:
Ah, yeah you're correct, it should be (also in other places). No shortcuts :P